src/plugins/hidsim/gui/vmi_win_gui_parser.cpp

```diff
@@ -0,0 +1,1040 @@
+        result->y1 = y1;
+        result->w = x1 - x0;
+        result->h = y1 - y0;
+    }
+
+    return result;
+}
+
+void update_visibility_bitmask(char* map, size_t n, int scanline,
+    struct rect* r)
+{
+    size_t byte;
+    int bit_idx;
+    unsigned int bit;
+
+    for (int y = r->y0; y < r->y1; y++)
+    {
+        for (int x = r->x0; x < r->x1; x++)
+        {
+            byte = (y * scanline + x)/8;
+
+            if (x < 0 || byte >= n )
+                continue;
+
+            /* Parts of a wnd can be outside of the desktop's frame */
+            bit_idx = x % 8;
+            bit = 0x80 >> bit_idx;
+            map[byte] |= bit;
+        }
+    }
+}
+
+
+/*
+ * Determine, if windows is visible;
+ * important since invisible windows might have visible children
+ */
+bool is_wnd_visible(vmi_instance_t vmi, vmi_pid_t pid, addr_t wnd)
+{
+    uint32_t style = 0;
+
+    if (VMI_FAILURE == vmi_read_32_va(vmi, wnd + symbol_offsets.wnd_style, pid, (uint32_t*)&style))
+        return false;
+
+    if (style & WS_VISIBLE)
+        return true;
+
+    return false;
+}
+
+struct wnd* construct_wnd_container(vmi_instance_t vmi, vmi_pid_t pid, addr_t win)
+{
+    status_t ret = VMI_FAILURE;
+
+    uint32_t x0 = 0;
+    uint32_t x1 = 0;
+    uint32_t y0 = 0;
+    uint32_t y1 = 0;
+
+    uint32_t rx0 = 0;
+    uint32_t rx1 = 0;
+    uint32_t ry0 = 0;
+    uint32_t ry1 = 0;
+
+    uint32_t style = 0;
+    uint32_t exstyle = 0;
+    wchar_t* wn = NULL;
+
+    ret = vmi_read_32_va(vmi, win + symbol_offsets.rc_wnd_offset + symbol_offsets.rc_left_offset, pid, (uint32_t*)&x0);
+    ret = vmi_read_32_va(vmi, win + symbol_offsets.rc_wnd_offset + symbol_offsets.rc_right_offset, pid, (uint32_t*)&x1);
+    ret = vmi_read_32_va(vmi, win + symbol_offsets.rc_wnd_offset + symbol_offsets.rc_top_offset, pid, (uint32_t*)&y0);
+    ret = vmi_read_32_va(vmi, win + symbol_offsets.rc_wnd_offset + symbol_offsets.rc_bottom_offset, pid, (uint32_t*)&y1);
+
+    /* Determine, if windows is visible */
+    ret = vmi_read_32_va(vmi, win + symbol_offsets.wnd_style, pid, (uint32_t*)&style);
+
+    /* Determine extended style attributes */
+    ret = vmi_read_32_va(vmi, win + symbol_offsets.wnd_exstyle, pid, (uint32_t*)&exstyle);
+
+    /* Retrieves atom value */
+    addr_t pcls = 0;
+    ret = vmi_read_addr_va(vmi, win + symbol_offsets.pcls_offset, pid, &pcls);
+    uint16_t atom = 0;
+    ret = vmi_read_16_va(vmi, pcls + symbol_offsets.cls_atom_offset, pid, &atom);
+
+    ret = vmi_read_32_va(vmi, win + symbol_offsets.rc_client_offset + symbol_offsets.rc_left_offset, pid, (uint32_t*)&rx0);
+    ret = vmi_read_32_va(vmi, win + symbol_offsets.rc_client_offset + symbol_offsets.rc_right_offset, pid, (uint32_t*)&rx1);
+    ret = vmi_read_32_va(vmi, win + symbol_offsets.rc_client_offset + symbol_offsets.rc_top_offset, pid, (uint32_t*)&ry0);
+    ret = vmi_read_32_va(vmi, win + symbol_offsets.rc_client_offset + symbol_offsets.rc_bottom_offset, pid, (uint32_t*)&ry1);
```

Comment:
More redundant casts in this function.
And I don't really get why set the `ret` variable  if we're going to overwrite it anyway :thinking:

Reply:
> And I don't really get why set the ret variable if we're going to overwrite it anyway thinking

Thanks for catching this bug. Changed in 6c187990d3e5598875d99bc4ceada013e7b3c167